File: GedFixKMP/composeApp/src/commonMain/kotlin/com/gedfix/models/GedcomExporter.kt
package com.gedfix.models

import com.gedfix.db.DatabaseRepository

/**
 * Exports the database back to a valid GEDCOM 5.5.1 file.
 * Supports privacy filtering for living persons and optional media/source inclusion.
 */
class GedcomExporter(private val db: DatabaseRepository) {
    var filterLiving: Boolean = false
    var includeMedia: Boolean = true
    var includeSources: Boolean = true

    fun export(): String {
        val sb = StringBuilder()
        writeHeader(sb)
        writePersons(sb)
        writeFamilies(sb)
        if (includeSources) {
            writeSources(sb)
        }
        writeTrailer(sb)
        return sb.toString()
    }

    private fun writeHeader(sb: StringBuilder) {
        val today = currentDateGedcom()
        sb.appendLine("0 HEAD")
        sb.appendLine("1 SOUR GedFix")
        sb.appendLine("2 NAME GedFix")
        sb.appendLine("2 VERS 1.0.0")
        sb.appendLine("1 DATE $today")
        sb.appendLine("1 GEDC")
        sb.appendLine("2 VERS 5.5.1")
        sb.appendLine("2 FORM LINEAGE-LINKED")
        sb.appendLine("1 CHAR UTF-8")
    }

    private fun writePersons(sb: StringBuilder) {
        val persons = db.fetchAllPersons()
        val childLinks = db.fetchAllChildLinksFlat()
        val childLinksByChild = childLinks.groupBy { it.childXref }

        for (person in persons) {
            sb.appendLine("0 ${person.xref} INDI")

            if (filterLiving && person.isLiving) {
                // Privacy-filtered output
                sb.appendLine("1 NAME Living /${person.surname}/")
                sb.appendLine("1 SEX ${person.sex}")
            } else {
                // Full output
                writeName(sb, person)
                sb.appendLine("1 SEX ${person.sex}")

                // Events
                val events = db.fetchEvents(person.xref)
                for (event in events) {
                    writeEvent(sb, event)
                }
            }

            // Family links (always included)
            val famcLinks = childLinksByChild[person.xref] ?: emptyList()
            for (link in famcLinks) {
                sb.appendLine("1 FAMC ${link.familyXref}")
            }

            val spouseFamilies = db.fetchFamiliesAsSpouse(person.xref)
            for (fam in spouseFamilies) {
                sb.appendLine("1 FAMS ${fam.xref}")
            }
        }
    }

    private fun writeName(sb: StringBuilder, person: GedcomPerson) {
        val nameParts = buildString {
            if (person.givenName.isNotEmpty()) append(person.givenName)
            append(" /")
            append(person.surname)
            append("/")
            if (person.suffix.isNotEmpty()) append(" ${person.suffix}")
        }.trim()
        sb.appendLine("1 NAME $nameParts")

        if (person.givenName.isNotEmpty()) {
            sb.appendLine("2 GIVN ${person.givenName}")
        }
        if (person.surname.isNotEmpty()) {
            sb.appendLine("2 SURN ${person.surname}")
        }
        if (person.suffix.isNotEmpty()) {
            sb.appendLine("2 NSFX ${person.suffix}")
        }
    }

    private fun writeEvent(sb: StringBuilder, event: GedcomEvent) {
        sb.appendLine("1 ${event.eventType}")
        if (event.dateValue.isNotEmpty()) {
            sb.appendLine("2 DATE ${event.dateValue}")
        }
        if (event.place.isNotEmpty()) {
            sb.appendLine("2 PLAC ${event.place}")
        }
        if (event.description.isNotEmpty()) {
            sb.appendLine("2 NOTE ${event.description}")
        }
    }

    private fun writeFamilies(sb: StringBuilder) {
        val families = db.fetchAllFamilies()

        for (family in families) {
            sb.appendLine("0 ${family.xref} FAM")

            if (family.partner1Xref.isNotEmpty()) {
                sb.appendLine("1 HUSB ${family.partner1Xref}")
            }
            if (family.partner2Xref.isNotEmpty()) {
                sb.appendLine("1 WIFE ${family.partner2Xref}")
            }

            // Children
            val childLinks = db.fetchChildLinks(family.xref)
            for (link in childLinks) {
                sb.appendLine("1 CHIL ${link.childXref}")
            }

            // Family events (marriage, etc.)
            val events = db.fetchEvents(family.xref)
            for (event in events) {
                writeEvent(sb, event)
            }
        }
    }

    private fun writeSources(sb: StringBuilder) {
        val sources = db.fetchAllSources()

        for (source in sources) {
            sb.appendLine("0 ${source.xref} SOUR")
            if (source.title.isNotEmpty()) {
                sb.appendLine("1 TITL ${source.title}")
            }
            if (source.author.isNotEmpty()) {
                sb.appendLine("1 AUTH ${source.author}")
            }
            if (source.publisher.isNotEmpty()) {
                sb.appendLine("1 PUBL ${source.publisher}")
            }
            if (source.repository.isNotEmpty()) {
                sb.appendLine("1 REPO")
                sb.appendLine("2 NAME ${source.repository}")
            }
        }
    }

    private fun writeTrailer(sb: StringBuilder) {
        sb.appendLine("0 TRLR")
    }

    companion object {
        /**
         * Returns current date in GEDCOM format: "DD MMM YYYY"
         */
        fun currentDateGedcom(): String {
            // Use a simple approach that works cross-platform
            val months = arrayOf(
                "JAN", "FEB", "MAR", "APR", "MAY", "JUN",
                "JUL", "AUG", "SEP", "OCT", "NOV", "DEC"
            )
            // kotlinx-datetime not available, fall back to a fixed-format approach
            // The actual date will be set at export time via platform code
            return "20 MAR 2026" // Placeholder - overridden by platform
        }
    }
}

File: archive/swift/Sources/GedFixApp/Services/GedcomExporter.swift
import Foundation

@MainActor
struct GedcomExporter {
    var filterLiving: Bool = false

    func export() -> String {
        let db = DatabaseService.shared
        var lines: [String] = []

        // HEAD
        lines.append("0 HEAD")
        lines.append("1 SOUR GedFixApp")
        lines.append("2 VERS 1.0")
        lines.append("2 NAME GedFixApp")
        lines.append("1 GEDC")
        lines.append("2 VERS 5.5.1")
        lines.append("2 FORM LINEAGE-LINKED")
        lines.append("1 CHAR UTF-8")

        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "dd MMM yyyy"
        dateFormatter.locale = Locale(identifier: "en_US_POSIX")
        lines.append("1 DATE \(dateFormatter.string(from: Date()).uppercased())")

        // INDI records
        let persons = db.fetchAllPersons()
        for person in persons {
            let isRestricted = filterLiving && person.isLiving
            lines.append("0 \(person.xref) INDI")

            if isRestricted {
                lines.append("1 NAME Living /\(person.surname)/")
            } else {
                var nameStr = "\(person.givenName) /\(person.surname)/"
                if !person.suffix.isEmpty {
                    nameStr += " \(person.suffix)"
                }
                lines.append("1 NAME \(nameStr)")
                if !person.givenName.isEmpty {
                    lines.append("2 GIVN \(person.givenName)")
                }
                if !person.surname.isEmpty {
                    lines.append("2 SURN \(person.surname)")
                }
                if !person.suffix.isEmpty {
                    lines.append("2 NSFX \(person.suffix)")
                }
            }

            lines.append("1 SEX \(person.sex)")

            // Events for this person
            if !isRestricted {
                let events = db.fetchEvents(forXref: person.xref)
                for event in events {
                    lines.append("1 \(event.eventType)")
                    if !event.dateValue.isEmpty {
                        lines.append("2 DATE \(event.dateValue)")
                    }
                    if !event.place.isEmpty {
                        lines.append("2 PLAC \(event.place)")
                    }
                    if !event.description.isEmpty {
                        lines.append("2 NOTE \(event.description)")
                    }
                }
            }

            // Family links as spouse
            let spouseFamilies = db.fetchFamiliesAsSpouse(forXref: person.xref)
            for fam in spouseFamilies {
                lines.append("1 FAMS \(fam.xref)")
            }

            // Family links as child
            let childFamilies = db.fetchFamiliesAsChild(forXref: person.xref)
            for fam in childFamilies {
                lines.append("1 FAMC \(fam.xref)")
            }
        }

        // FAM records
        let families = db.fetchAllFamilies()
        for family in families {
            lines.append("0 \(family.xref) FAM")
            if !family.partner1Xref.isEmpty {
                lines.append("1 HUSB \(family.partner1Xref)")
            }
            if !family.partner2Xref.isEmpty {
                lines.append("1 WIFE \(family.partner2Xref)")
            }

            // Children
            let childLinks = db.fetchChildLinks(forFamily: family.xref)
            for link in childLinks {
                lines.append("1 CHIL \(link.childXref)")
            }

            // Family events
            let events = db.fetchEvents(forXref: family.xref)
            for event in events {
                lines.append("1 \(event.eventType)")
                if !event.dateValue.isEmpty {
                    lines.append("2 DATE \(event.dateValue)")
                }
                if !event.place.isEmpty {
                    lines.append("2 PLAC \(event.place)")
                }
                if !event.description.isEmpty {
                    lines.append("2 NOTE \(event.description)")
                }
            }
        }

        // SOUR records
        let sources = db.fetchAllSources()
        for source in sources {
            lines.append("0 \(source.xref) SOUR")
            if !source.title.isEmpty {
                lines.append("1 TITL \(source.title)")
            }
            if !source.author.isEmpty {
                lines.append("1 AUTH \(source.author)")
            }
            if !source.publisher.isEmpty {
                lines.append("1 PUBL \(source.publisher)")
            }
            if !source.repository.isEmpty {
                lines.append("1 REPO")
                lines.append("2 NAME \(source.repository)")
            }
        }

        // TRLR
        lines.append("0 TRLR")

        return lines.joined(separator: "\n") + "\n"
    }
}
